### PR TITLE
Fix CVE-2018-1000544 

### PR DESCRIPTION
**Pull request by bdewater**
Attempt to fix https://github.com/rubyzip/rubyzip/issues/369

Not sure if there's a better way to handle Errno::ENOENT in the symlink case, open to suggstions.

### CODE REVIEW

File: lib/zip/entry.rb [hunk 1]
```diff
@@ -147,14 +147,21 @@ def next_header_offset #:nodoc:all
...
-
-      if @name.squeeze('/') =~ /\.{2}(?:\/|\z)/
+    def extract(dest_path = nil, &block)
+      if dest_path.nil? && Pathname.new(@name).absolute?
+        puts "WARNING: skipped absolute path in #{@name}"
+        return self
+      elsif @name.squeeze('/') =~ /\.{2}(?:\/|\z)/
         puts "WARNING: skipped \"../\" path component(s) in #{@name}"
         return self
+      elsif symlink? && get_input_stream.read =~ %r{../..}
```
Comment (JackMc) [inline lib/zip/entry.rb]: Could you add a test that this still works for non-malicious symlinks? It doesn't appear there is one.

The comment on `get_input_stream`'s definition seems to also say that it doesn't work too great for symlinks. We should make sure we're not adding any undefined behaviour: https://github.com/rubyzip/rubyzip/blob/master/lib/zip/entry.rb#L484

Comment (tuzovakaoff) [reply lib/zip/entry.rb]: Symlink could also links to absolute path
You can find examples of such zip files in https://github.com/jwilk/path-traversal-samples

**Comment by zeha** [comment]
Given this doesn't look too broken, can we please have that merged and a new release stamped with it? Thanks :-)

File: lib/zip/entry.rb [hunk 2]
```diff
@@ -147,14 +147,21 @@ def next_header_offset #:nodoc:all
...
     # Extracts entry to file dest_path (defaults to @name).
-    def extract(dest_path = @name, &block)
-      block ||= proc { ::Zip.on_exists_proc }
-
-      if @name.squeeze('/') =~ /\.{2}(?:\/|\z)/
+    def extract(dest_path = nil, &block)
+      if dest_path.nil? && Pathname.new(@name).absolute?
+        puts "WARNING: skipped absolute path in #{@name}"
+        return self
+      elsif @name.squeeze('/') =~ /\.{2}(?:\/|\z)/
```
Comment (zeha) [inline lib/zip/entry.rb]: Should this not also gain the `dest_path.nil?` check?

**Comment by zeha** [comment]
Example in README calls .extract() with the name from the zipfile, leading to the same vulnerability in code that gets copy-pasted from the README.

**Comment by drn** [comment]
Our team would love to see this merged and released as well. This has started to show up in our automated vulnerability analysis (https://hakiri.io/technologies/rubyzip/issues/1e6c19ff65e713) and our builds are red as a result.

Happy to help get this over the line if any additional work is needed. Please let me know!

**Comment by bdewater** [comment]
I can address feedback to this PR - @simonoff have you had a chance to look at this?

Right now do not have the bandwidth available to look at/add tests for all the examples in https://github.com/jwilk/path-traversal-samples - so if somebody feels like taking a crack at it you're most welcome 🙂

**Comment by simonoff** [comment]
Tests is failed. Need to have a green build to make it merged.

**Comment by bdewater** [comment]
@simonoff like I [mentioned](https://github.com/rubyzip/rubyzip/pull/371#issuecomment-401630960) the test failures seem unrelated - can you try to rebuild?

**Comment by reedloden** [comment]
This was just added to [ruby-advisory-db](https://github.com/rubysec/ruby-advisory-db/pull/346), so would love for this PR to be finalized, landed, and released.

**Comment by vipulnsward** [comment]
I looked at why the jruby builds were failing. Its after an upgrade to 9.2 affecting the crc32 checksum in Encryption tests. 
We can ignore failures on JRuby, since its a difference in MRI and JRuby implementation for now.

**Comment by bdewater** [comment]
Thanks @vipulnsward - I've pulled your commit in to get this PR green.

**Comment by jdleesmiller** [comment]
Based on https://github.com/rubyzip/rubyzip/pull/371#discussion_r199429562 and https://github.com/rubyzip/rubyzip/pull/371#issuecomment-414782527 I added tests for https://github.com/jwilk/path-traversal-samples

Tests are in: https://github.com/jdleesmiller/rubyzip/commit/9c468f30f38d09451e5a65edfff277cfe381fd49

Good news: most pass.

Bad news: two don't pass: `test_directory_symlink` and `test_two_directory_symlinks_b`:

```
  1) Failure:
PathTraversalTest#test_two_directory_symlinks_b [/Users/john/ex/rubyzip/test/path_traversal_test.rb:85]:
Expected false to be truthy.

  2) Failure:
PathTraversalTest#test_directory_symlink [/Users/john/ex/rubyzip/test/path_traversal_test.rb:65]:
Expected false to be truthy.
```

I have not yet attempted to fix the bugs, and I'm not sure I know enough about this library to do that, but I thought I should at least report the test results.

**Pull request merged**